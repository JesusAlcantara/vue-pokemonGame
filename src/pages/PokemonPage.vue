<template>
    <h1 v-if="!pokemon">Espere por favor...</h1>

    <div v-else>
        <h1>Cuál es este pokémon?</h1>
        <pokemon-picture 
            :pokemonId="pokemon.id"
            :showPokemon="showPokemon"
        />
        <pokemon-options 
            :pokemons="pokemonArr"
            @selectionPokemon="checkAnswer"
        />
        
        <template v-if="showAnswer">
            <h2 class="fade-in">
                {{ message }}
            </h2>
            <button @click="newGame()">
                Nuevo Juego
            </button>
        </template>
    </div>
</template>

<script>
import PokemonOptions from '@/components/PokemonOptions.vue';
import PokemonPicture from '@/components/PokemonPicture.vue';
import getPokemonOptions from '@/helpers/getPokemonOptions';

export default {
    components: {
        PokemonPicture,
        PokemonOptions
    },
    data() {
        return {
            pokemonArr: [],
            pokemon: null,
            showPokemon: false,
            showAnswer: false,
            message: ''
        }
    },
    methods: {
        async mixPokemonArray () {
            this.pokemonArr = await getPokemonOptions();
            const rndInt = Math.floor(Math.random() * 4);
            this.pokemon = this.pokemonArr[rndInt];
        },
        checkAnswer (selectedId) {
            this.showPokemon = true;
            this.showAnswer = true;
            if ( selectedId === this.pokemon.id ) {
                this.message = `¡Correcto es ${this.capitalized(this.pokemon.name)}!`
            } else {
                this.message = `Opps, el pokémon era ${this.capitalized(this.pokemon.name)}`
            }
        },
        newGame () {
            this.showPokemon = false;
            this.showAnswer = false;
            this.pokemonArr = [];
            this.pokemon = null;
            this.mixPokemonArray();
        },
        capitalized(name) {
            const capitalizedFirst = name[0].toUpperCase();
            const rest = name.slice(1);
            return capitalizedFirst + rest;
        }
    },
    mounted() {
        this.mixPokemonArray();
    }
}
</script>