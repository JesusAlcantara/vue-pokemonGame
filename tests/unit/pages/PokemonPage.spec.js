import PokemonPage from '@/pages/PokemonPage';
import { shallowMount, mount } from '@vue/test-utils';
import { mockPokemons } from '../mocks/pokemons.mock';

describe('PokemonPage page', () => {

    let wrapper;
    let mixPokemonArraySpy;
    beforeEach(() => {
        mixPokemonArraySpy = jest.spyOn(PokemonPage.methods, 'mixPokemonArray');
        capitalizedSpy = jest.spyOn(PokemonPage.methods, 'capitalized');
        wrapper = shallowMount(PokemonPage);
    })

    describe('Render cases', () => {
        it('should match with snapshot', () => {
            expect(wrapper.html()).toMatchSnapshot();
        })

        it('should match with snapshot when pokemons load', () => {
            const wrapper = shallowMount(PokemonPage, {
                data() {
                    return {
                        pokemonArr: mockPokemons,
                        pokemon: mockPokemons[0],
                        showPokemon: false,
                        showAnswer: false,
                        message: ''
                    }
                }
            });
            expect(wrapper.html()).toMatchSnapshot();
        })
    
        it('should show PokemonPicture and PokemonOptions', () => {
            const wrapper = shallowMount(PokemonPage, {
                data() {
                    return {
                        pokemonArr: mockPokemons,
                        pokemon: mockPokemons[0],
                        showPokemon: false,
                        showAnswer: false,
                        message: ''
                    }
                }
            });
            const picture = wrapper.find('pokemon-picture-stub');
            const options = wrapper.find('pokemon-options-stub');
            expect(picture.exists()).toBe(true);
            expect(options.exists()).toBe(true);
            expect(picture.attributes('pokemonid')).toBe('5');
            expect(options.attributes('pokemons')).toBeTruthy();
        })
    })

    describe('Methods', () => {
        it('should call "mixPokemonArray" in mount', () => {
            expect(mixPokemonArraySpy).toHaveBeenCalled();
        })

        it('should call checkAnswer', async () => {
            const wrapper = shallowMount(PokemonPage, {
                data() {
                    return {
                        pokemonArr: mockPokemons,
                        pokemon: mockPokemons[0],
                        showPokemon: false,
                        showAnswer: false,
                        message: ''
                    }
                }
            });
            await wrapper.vm.checkAnswer(5);
            expect(wrapper.find('h2').exists()).toBe(true);
            expect(wrapper.vm.showPokemon).toBe(true);
            expect(wrapper.find('h2').text()).toBe(`¡Correcto es ${capitalizedSpy(mockPokemons[0].name)}!`);
            await wrapper.vm.checkAnswer(10);
            expect(wrapper.vm.message).toBe(`Opps, el pokémon era ${capitalizedSpy(mockPokemons[0].name)}`);
        })
    })
})