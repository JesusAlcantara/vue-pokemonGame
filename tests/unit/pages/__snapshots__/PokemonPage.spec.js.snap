// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage page Render cases should match with snapshot 1`] = `<h1>Espere por favor...</h1>`;

exports[`PokemonPage page Render cases should match with snapshot when pokemons load 1`] = `
<div>
  <h1>Cuál es este pokémon?</h1>
  <pokemon-picture-stub pokemonid="5" showpokemon="false"></pokemon-picture-stub>
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--v-if-->
</div>
`;
